src/lib.rs

```diff
@@ -21,56 +21,72 @@
 // initially, and then 11 cycles per outer loop. We ignore
 // all that.
 
-/// Internal function to implement a variable busy-wait loop.
+///delay for N milliseconds
 /// # Arguments
-/// * 'count' - an i32, the number of times to cycle the loop.
-pub fn delay(count: u32) {
-    // Our asm busy-wait takes a 16 bit word as an argument,
-    // so the max number of loops is 2^16
-    let outer_count = count / 65536;
-    let last_count = ((count % 65536)+1) as u16;
-    for _ in 0..outer_count {
-        // Each loop through should be 4 cycles.
-        unsafe {llvm_asm!("1: sbiw $0,1
-                      brne 1b"
-                     :
-                     : "w" (0)
-                     :
-                     :)}
-    }
-    unsafe {llvm_asm!("1: sbiw $0,1
-                      brne 1b"
-                 :
-                 : "w" (last_count)
-                 :
-                 :)}
-}
-
-///delay for N miliseconds
-/// # Arguments
-/// * 'ms' - an u32, number of milliseconds to busy-wait
+/// * 'ms' - a u32, number of milliseconds to busy-wait
+#[inline(always)]
 pub fn delay_ms(ms: u32) {
-    // microseconds
-    let us = ms * 1000;
-    delay_us(us);
+    let ticks: u64 = (u64::from(avr_config::CPU_FREQUENCY_HZ) * u64::from(ms)) / 4_000;
```

Comment:
I don't know if it depends on the compile time evaluation, more like it hopes it does it? My naive understanding is compile time evaluation wouldn't be guaranteed in the C version.

The instruction set does have some float multiplication, but not float divisions. I don't know how costly floats are maybe its not that bad? This was written by Atmel, I would hope they know a thing or two about their CPUs.

Float would have no minute limit while retaining high precision for small values. I will investigate and report here.

Reply:
This is from `<util/delay.h>` (the file defining `_delay_ms` and `_delay_us`) included with Microchip Studio:

>  The functions in this header file are wrappers around the basic busy-wait functions from <util/delay_basic.h>.  They are meant as convenience functions where actual time values can be specified rather than a number of cycles to wait for.  The idea behind is that compile-time constant expressions will be eliminated by compiler optimization so floating-point expressions can be used to calculate the number of delay cycles needed based on the CPU frequency passed by the macro F_CPU.
>
> In order for these functions to work as intended, compiler optimizations *must* be enabled, and the delay time *must* be an expression that is a known constant at compile-time.  If these requirements are not met, the resulting delay will be much longer (and basically unpredictable), and  applications that otherwise do not use floating-point calculations will experience severe code bloat by the floating-point library routines linked into the application.

Reply:
Since AVR doesn't have hardware division at all, I think it's basically impossible to make delay_ms and delay_us work with values that aren't known at compile time without a ton of overhead. At the very least, I'd say it isn't a priority for this crate.

Reply:
The division is not needed at runtime in my example. But it still requires 16bit substraction and multiplication.

But note that even if you call delay with a constant value. The compiler might still move the computation at runtime to reduce code bloat. I encountered this on my last project (attiny24a, 2048 bytes of flash). I used delays to output patterns on a pin, which helped me synchronize debugging on an oscilloscope. The trick was to use only two different delay constant. The compiler was happy to output only two functions.

Reply:
example, with opt-level=z:
```rust
#[no_mangle]
pub unsafe extern "C" fn main() {
    busy_loop_ms(66);
    busy_loop_ms(67);
    busy_loop_ms(68);
    loop {}
}
```

Produces:
```asm
main:
	ldi	r24, 66
	ldi	r25, 0
	rcall	_ZN12attiny24a_uc3mcu5delay12busy_loop_ms17hdb07dc8da5567d03E
	ldi	r24, 67
	ldi	r25, 0
	rcall	_ZN12attiny24a_uc3mcu5delay12busy_loop_ms17hdb07dc8da5567d03E
	ldi	r24, 68
	ldi	r25, 0
	rcall	_ZN12attiny24a_uc3mcu5delay12busy_loop_ms17hdb07dc8da5567d03E
.LBB3_1:
	rjmp	.LBB3_1

_ZN12attiny24a_uc3mcu5delay12busy_loop_ms17hdb07dc8da5567d03E:
	ldi	r18, 0
	ldi	r19, 250
.LBB0_1:
	cpi	r24, 33
	cpc	r25, r1
	brlo	.LBB0_3
	movw	r30, r18
	;APP
.Ltmp0:
	sbiw	r30, 1
	brne	.Ltmp0
	;NO_APP
	sbiw	r24, 32
	rjmp	.LBB0_1
.LBB0_3:
	cpi	r24, 0
	cpc	r25, r1
	breq	.LBB0_5
	ldi	r22, 208
	ldi	r23, 7
	rcall	__mulhi3
	;APP
.Ltmp1:
	sbiw	r24, 1
	brne	.Ltmp1
	;NO_APP
.LBB0_5:
	ret
```

Reply:
How are you getting that assembly output? I've been using avr-objdump on the .elf output file, and the formatting is much less nice

Reply:
`cargo rustc $(FLAGS) --target=avr-attiny24a.json -- --emit asm`